### DISCUSSION TITLE
When exactly does the __MISE_WATCH or __MISE_DIFF env variable change?

### DISCUSSION THREAD

**Discussion by ffittschen** [jdx/mise#2213] [Q&A]
Hi all,
I'm curious when exactly the `__MISE_WATCH` or `__MISE_DIFF` environment variable changes. 

When tracking down some issues with the Swift Package Manager that appeared after installing `mise`, I discovered that SPM actually uses the environment as input for their cache key (see [implementation](https://github.com/apple/swift-package-manager/blob/e9399c20bab8c53b77dd6ed1529f68eb47dc4daf/Sources/PackageLoading/ManifestLoader.swift#L565-L573), [discussion](https://forums.swift.org/t/rebuild-times-in-new-shell-significantly-slower/68054/2)). In Swift 5.10, they actually [added a filter](https://github.com/apple/swift-package-manager/pull/7056) to remove some env variables, which change very often, which led me to debug my env variables.

I ran `printenv` twice while waiting 1 minute between the invocations and observed that the `__MISE_WATCH` env variable changed, without me touching anything on the system. That's why I'm asking the question.

**Comment by jdx** [comment]
that is actually a very difficult question to answer—I have no idea. The code doesn't "change" it, it generates it every time the prompt is displayed. You'd have to trace the code to see.

If it's a problem you could just not use `mise activate`.

**Reply by ffittschen** [reply]
I traced the code to the [config.watch_files()](https://github.com/jdx/mise/blob/ef25cb1a419a39de42a2839b4e97c56f19bfe569/src/config/mod.rs#L396-L406) function.

As far as I can see, it takes the following as input:
- all config files paths
- all resolved env_files paths
- all resolved env_scripts paths
- the global MISE_ENV_FILE path

As this still a bit vague, I just looked into deserializing the `__MISE_WATCH` value manually[^1], resulting in this structure

```
files:
  /Users/<username>/.config/mise/config.toml: <modified timestamp>
  /Users/<username>/.local/share/mise: <modified timestamp>
  /Users/<username>/<project-dir>/.mise.toml: <modified timestamp>
  /Users/<username>/<project-dir>/Gemfile: <modified timestamp>
env_var_hash: 37a317d896eaf472
```

What I do not understand is why the value of `__MISE_WATCH` changes if I don't touch any of those files mentioned above. The `env_var_hash` should also stay the same as the only env var starting with `MISE_` is `MISE_SHELL`, which is static.

From running the command again after waiting a while, I can see that the value mise sets as the modified timestamp of the `/Users/<username>/.local/share/mise` directory changes, but when running `stat /Users/<username>/.local/share/mise`, the mtime value stays the same... 

Do you have an idea what could modify the `/Users/<username>/.local/share/mise` directory?

[^1]: `echo "$__MISE_WATCH" | base64 --decode | pigz -d -z --fast`

**Reply by jdx** [reply]
yeah, it gets touched right here: https://github.com/jdx/mise/blob/ef25cb1a419a39de42a2839b4e97c56f19bfe569/src/cli/activate.rs#L61 but are you actually running `mise activate` somewhere?

it also gets touched here: https://github.com/jdx/mise/blob/ef25cb1a419a39de42a2839b4e97c56f19bfe569/src/cli/activate.rs#L61 but that is when installing tools

**Reply by jdx** [reply]
might be worth trying removing those lines and seeing if that stops the mtime from changing

**Comment by Marukome0743** [comment]
The implementation has changed since this discussion was opened, so the original `__MISE_WATCH` behavior no longer applies to current mise versions.

[PR #3532](https://github.com/jdx/mise/pull/3532), included in mise v2024.12.10, replaced the old `__MISE_WATCH` state with the structured `__MISE_SESSION` mechanism. Current mise no longer exports `__MISE_WATCH`.

`__MISE_DIFF` is still used to record the environment changes managed by mise so that they can be reversed or updated later. It is not regenerated with a new time-dependent value on every prompt when nothing has changed.

I verified this on current main with Bash activation: `__MISE_WATCH` was absent, and `__MISE_DIFF` remained byte-for-byte identical across consecutive prompt-hook executions with no config, directory, or managed-environment changes.

Therefore, the Swift Package Manager cache-key churn originally observed from `__MISE_WATCH` should no longer occur on current mise releases. `__MISE_SESSION` or `__MISE_DIFF` may still change when mise actually detects a relevant config, directory, tool, or managed-environment change.

*AI-assisted — Tool: Codex; model: OpenAI/GPT-5; version: unavailable.*
